server/src/main/java/org/apache/iotdb/db/mpp/execution/DataDriver.java

```diff
@@ -0,0 +1,324 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.mpp.execution;
+
+import org.apache.iotdb.db.engine.querycontext.QueryDataSource;
+import org.apache.iotdb.db.engine.storagegroup.TsFileResource;
+import org.apache.iotdb.db.engine.storagegroup.VirtualStorageGroupProcessor;
+import org.apache.iotdb.db.exception.query.QueryProcessException;
+import org.apache.iotdb.db.metadata.idtable.IDTable;
+import org.apache.iotdb.db.metadata.path.PartialPath;
+import org.apache.iotdb.db.mpp.buffer.ISinkHandle;
+import org.apache.iotdb.db.mpp.common.FragmentInstanceId;
+import org.apache.iotdb.db.mpp.operator.Operator;
+import org.apache.iotdb.db.mpp.operator.source.SourceOperator;
+import org.apache.iotdb.db.query.control.FileReaderManager;
+import org.apache.iotdb.tsfile.read.common.block.TsBlock;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.units.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static org.apache.iotdb.db.mpp.operator.Operator.NOT_BLOCKED;
+
+@NotThreadSafe
+public class DataDriver implements ExecFragmentInstance {
+
+  private static final Logger logger = LoggerFactory.getLogger(DataDriver.class);
+
+  private final Operator root;
+  private final ISinkHandle sinkHandle;
+  private final DataDriverContext driverContext;
+
+  private boolean init;
+  private boolean closed;
+
+  /** closed tsfile used in this fragment instance */
+  private Set<TsFileResource> closedFilePaths;
+  /** unClosed tsfile used in this fragment instance */
+  private Set<TsFileResource> unClosedFilePaths;
+
+  private final AtomicReference<SettableFuture<Void>> driverBlockedFuture = new AtomicReference<>();
+
+  public DataDriver(Operator root, ISinkHandle sinkHandle, DataDriverContext driverContext) {
+    this.root = root;
+    this.sinkHandle = sinkHandle;
+    this.driverContext = driverContext;
+    this.closedFilePaths = new HashSet<>();
+    this.unClosedFilePaths = new HashSet<>();
+    // initially the driverBlockedFuture is not blocked (it is completed)
+    SettableFuture<Void> future = SettableFuture.create();
+    future.set(null);
+    driverBlockedFuture.set(future);
+  }
+
+  @Override
+  public boolean isFinished() {
+    try {
+      boolean isFinished =
+          closed || (driverBlockedFuture.get().isDone() && root != null && root.isFinished());
```

Comment:
1. 从实现上看，finished 语义表示的是是否需要继续被调度执行吧？ExecFragmentInstance#isFinished 的 javadoc可能有点不准确？
2. driverBlockedFuture 是否 complete 和 finished 似乎没有关系

Reply:
Yes, here just want to make sure that if driverBlockedFuture is not done, isFinished method should never return true, otherwise it will have some problems.

Reply:
> 1. 从实现上看，finished 语义表示的是是否需要继续被调度执行吧？ExecFragmentInstance#isFinished 的 javadoc可能有点不准确？

Actually, I think there is a causal relationship between these two. This instance don't need to be scheduled because of this instance has no more data to process.

Reply:
> Actually, I think there is a causal relationship between these two. This instance don't need to be scheduled because of this instance has no more data to process.

An instance terminated due to failure also shouldn't be scheduled for execution anymore.

Reply:
That's true, i will fix that